### PR TITLE
feat: built-in fluent integration

### LINKED ISSUE #28: Narrow interfaces are not assignable to TranslationContext
`ctx.t` accepts `TranslationContext` that is defined as `Record<string, FluentValue>`
However, if you try to pass a narrower type (like `{a: string}`), you get an error

### LINKED ISSUE #29: Fluent dependencies are bundled on Node instead of being added to dependencies
image

### PR DESCRIPTION
**Pull request by dcdunkan**
Makes the Fluent class taken from [better_fluent](https://deno.land/x/better_fluent) built-in.

There are several reasons for this change:
- I'll make a PR for https://t.me/grammyjs/103520 next (I already have a implementation). For that we need a custom syntax for keys, and a custom way to manage the bundles. Its not something that I currently want to add to better_fluent. This PQ makes it easy to implement that feature.
- better_fluent can handle multiple resources for the same locale, which is not a feature being used in i18n. So, this fluent has less code, and makes it a little faster.
- Closes #29 because only [x/fluent](https://deno.land/x/fluent) is imported and that can be mapped to [@fluent/bundle](https://npmjs.com/package/@fluent/bundle) and [@fluent/langneg](https://npmjs.com/package/@fluent/langneg) in the dnt configuration. So, no transforming is required at all for the dependencies.
- Closes #28 as I've made the fixes and changes of #30 in this PQ inside the Fluent class as well.
- And something not so important: Don't have to depend on my repository, So, even when I'm unavailable [others can make the needed changes that's useful for i18n](https://github.com/grammyjs/i18n/pull/30#discussion_r1003499778) (I was and I'll be inactive for few months)

I don't think this has breaking changes in the point-of-view of i18n, because only the #30 changes are added in to `I18n`.

Feedback welcome.

### CODE REVIEW

File: deno.jsonc [hunk 1]
```diff
@@ -19,5 +19,6 @@
     "files": {
       "exclude": ["./out/"]
     }
-  }
+  },
+  "lock": false
```
Comment (rojvv) [inline deno.jsonc]: Move this up?

File: deno.jsonc [hunk 2]
```diff
@@ -12,7 +13,7 @@
   },
   "tasks": {
     "example": "cd examples && deno run --allow-net --allow-read deno.ts",
-    "test": "deno test --allow-read --allow-write",
+    "test": "deno test --allow-read --allow-write --allow-run",
```
Comment (rojvv) [inline deno.jsonc]:
```suggestion
    "test": "deno test --allow-read --allow-run --allow-write",
```

File: scripts/dnt.ts [hunk 3]
```diff
@@ -2,9 +2,9 @@ import {
   dirname,
   fromFileUrl,
   join,
-} from "https://deno.land/std@0.154.0/path/mod.ts";
+} from "https://deno.land/std@0.175.0/path/mod.ts";
```
Comment (rojvv) [inline scripts/dnt.ts]:
```suggestion
```

File: examples/deno.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import {
   Context,
   session,
   SessionFlavor,
-} from "https://deno.land/x/grammy@v1.11.0/mod.ts";
+} from "https://deno.land/x/grammy@v1.13.1/mod.ts";
```
Comment (rojvv) [inline examples/deno.ts]:
```suggestion
} from "https://deno.land/x/grammy@v1.14.1/mod.ts";
```

File: scripts/dnt.ts [hunk 5]
```diff
@@ -2,9 +2,9 @@ import {
...
```
Comment (rojvv) [inline scripts/dnt.ts]:
```suggestion
} from "https://deno.land/std@0.176.0/path/mod.ts";
```

File: src/deps.ts [hunk 6]
```diff
@@ -1,17 +1,15 @@
...
+export { negotiateLanguages } from "https://deno.land/x/fluent@v0.0.0/langneg/mod.ts";
 
 export {
   type Context,
   type HearsContext,
   type MiddlewareFn,
 } from "https://lib.deno.dev/x/grammy@1.x/mod.ts";
 
-export { extname, resolve } from "https://deno.land/std@0.154.0/path/mod.ts";
+export { extname, resolve } from "https://deno.land/std@0.175.0/path/mod.ts";
```
Comment (rojvv) [inline src/deps.ts]:
```suggestion
export { extname, resolve } from "https://deno.land/std@0.176.0/path/mod.ts";
```

File: src/i18n.ts [hunk 7]
```diff
@@ -111,19 +103,19 @@ export class I18n<C extends Context = Context> {
    * Gets a message by its key from the specified locale.
    * Alias of `translate` method.
```
Comment (rojvv) [inline src/i18n.ts]: Unrelated but would be cool

```suggestion
   * Alias of `translate`.
```

File: tests/deps.ts [hunk 8]
```diff
@@ -1,8 +1,10 @@
 export {
+  assert,
   assertEquals,
   assertNotEquals,
-} from "https://deno.land/std@0.153.0/testing/asserts.ts";
-export { join } from "https://deno.land/std@0.154.0/path/mod.ts";
+  assertStringIncludes,
+} from "https://deno.land/std@0.175.0/testing/asserts.ts";
+export { join } from "https://deno.land/std@0.175.0/path/mod.ts";
```
Comment (rojvv) [inline tests/deps.ts]:
```suggestion
} from "https://deno.land/std@0.176.0/testing/asserts.ts";
export { join } from "https://deno.land/std@0.176.0/path/mod.ts";
```

**Review by rojvv — APPROVED** [review]
LGTM.

**Review by Loskir — APPROVED** [review]

File: src/types.ts [hunk 9]
```diff
@@ -1,21 +1,40 @@
...
 
-export type TranslateFunction = (
+export type TranslationVariables<K extends string = string> = Record<
+  K,
+  FluentVariable
+>;
+export type TranslateFunction = <K extends string>(
   key: string,
-  context?: TranslationContext,
+  context?: TranslationVariables<K>,
```
Comment (EdJoPaTo) [inline src/types.ts]: changing the name of the type should also change the name of the variable. Guess this change is relevant in other places too. I think I suggested all of them but better check another time.
```suggestion
  variables?: TranslationVariables<K>,
```

File: tests/deps.node.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export function evalCommand() {
+  return ["node", "--eval"];
+}
+
+export function fluentImport() {
+  return 'const { Fluent, defaultWarningHandler } = require("./src/mod.js");';
```
Comment (EdJoPaTo) [inline tests/deps.node.ts]: This probably also involves DNT magic and I am mainly curious:
Isnt TypeScript using modules and `import` and resolving them to require when CommonJS is used?
Wouldnt it be better to use import here too? Or is there something else I am overlooking / dont know?

File: src/i18n.ts [hunk 11]
```diff
@@ -205,8 +185,17 @@ should either enable sessions or use `ctx.i18n.useLocale()` instead.",
...
+
+    ctx.translate = <K extends string>(
+      key: string,
+      translationContext?: TranslationVariables<K>,
+    ): string => {
+      return translate(key, {
+        ...globalTranslationContext?.(ctx),
+        ...translationContext,
+      });
+    };
```
Comment (EdJoPaTo) [inline src/i18n.ts]:
```suggestion
    ctx.translate = <K extends string>(
      key: string,
      translationVariables?: TranslationVariables<K>,
    ): string => {
      return translate(key, {
        ...globalTranslationContext?.(ctx),
        ...translationVariables,
      });
    };
```

File: src/i18n.ts [hunk 12]
```diff
@@ -109,21 +101,21 @@ export class I18n<C extends Context = Context> {
...
 
   /** Gets a message by its key from the specified locale. */
-  translate(
+  translate<K extends string>(
     locale: LocaleId,
     key: string,
-    context?: TranslationContext,
+    context?: TranslationVariables<K>,
   ): string {
     return this.fluent.translate(locale, key, context);
```
Comment (EdJoPaTo) [inline src/i18n.ts]:
```suggestion
    variables?: TranslationVariables<K>,
  ): string {
    return this.fluent.translate(locale, key, variables);
```

File: src/i18n.ts [hunk 13]
```diff
@@ -109,21 +101,21 @@ export class I18n<C extends Context = Context> {
...
+   * Alias of `translate`.
    */
-  t(
+  t<K extends string>(
     locale: LocaleId,
     key: string,
-    context?: TranslationContext,
+    context?: TranslationVariables<K>,
   ): string {
     return this.translate(locale, key, context);
```
Comment (EdJoPaTo) [inline src/i18n.ts]:
```suggestion
    variables?: TranslationVariables<K>,
  ): string {
    return this.translate(locale, key, variables);
```

**Review by EdJoPaTo — CHANGES_REQUESTED** [review]

File: tests/deps.node.ts (hunk 10 above)
Comment (dcdunkan) [reply tests/deps.node.ts]: This is for the tests that dnt runs after transforming the project to JS. So, we cannot use `import` syntax here, the tests will fail.

File: src/types.ts [hunk 14]
```diff
@@ -1,21 +1,40 @@
...
   | undefined
   | PromiseLike<LocaleId | undefined>;
 
-export type TranslateFunction = (
+export type TranslationVariables<K extends string = string> = Record<
+  K,
+  FluentVariable
+>;
+export type TranslateFunction = <K extends string>(
   key: string,
```
Comment (EdJoPaTo) [inline src/types.ts:36]: isn't the idea of `K` to be the key? wouldn't this be `key: K` then too?

Comment (dcdunkan) [reply src/types.ts:36]: Using `key: K` will cause another type error, no?

```ts
ctx.t("cart", { apples: ctx.session.apples });
//              ^!
```
```
Argument of type '{ apples: number; }' is not assignable to parameter of type 'TranslationVariables<"cart">'.
  Object literal may only specify known properties, and 'apples' does not exist in type 'TranslationVariables<"cart">'
```

I'm not really great with TS. I think we can have another similar generic type for key, but isn't just `string` type okay here? Could you maybe mention some usecase of this?

Comment (EdJoPaTo) [reply src/types.ts:36]: ah, right, key of variables, not identifier of the template to be used.

**Review by EdJoPaTo — APPROVED** [review]

**Comment by KnorpelSenf** [comment]
@dcdunkan or @EdJoPaTo feel free to release

**Pull request merged**